cs-minibot-platform-src/src/main/java/simulator/Simulator.java

```diff
@@ -223,4 +235,93 @@ public void importPhysicalObject(PhysicalObject pObj) {
         physicalObjectSet.add(pObj);
     }
 
+
+    public int[][] getOccupancyMatrix() {
+        return occupancyMatrix;
+    }
+
+    public void setOccupancyMatrix(int[][] om) {
+        occupancyMatrix = om;
+    }
+
+    public ArrayList<ArrayList<Integer>> getWayPoints(int[][]om) {
+        ArrayList<ArrayList<Integer>> points = new ArrayList<ArrayList<Integer>>();
+        for(int i = 1; i < om.length-1; i++) {
+            for(int j = 1; j < om[0].length-1; j++) {
+                if(om[i][j] == 1 &&
+                        ( (om[i-1][j] == 1 && om[i][j+1] == 1) || (om[i][j+1] == 1 && om[i+1][j] == 1)))
+                {
+                    points.get(0).add(i);
+                    points.get(1).add(j);
+                }
+            }
+        }
+        return points;
+    }
+
+
+    public void generateOccupancyMatrix(int occupancyMatrixHeight, int occupancyMatrixWidth, float occupancyMatrixBoxSize) {
+        int[][] om = new int[occupancyMatrixHeight][occupancyMatrixWidth];
+        //System.out.println("The occupancy matrix box size is " + occupancyMatrixBoxSize);
+        for(int i = 0; i < occupancyMatrixHeight; i++){
+            for(int j = 0; j < occupancyMatrixWidth; j++) {
+                float lower_x = (float)(i*occupancyMatrixBoxSize);
+                float lower_y = (float)(j*occupancyMatrixBoxSize);
+                Vec2 lowerVertex = new Vec2(lower_x, lower_y);
+                Vec2 upperVertex = new Vec2(lower_x + occupancyMatrixBoxSize, lower_y + occupancyMatrixBoxSize);
+                AABB currentSquare = new AABB(lowerVertex, upperVertex);
+                final Vec2 testpoint1 = new Vec2(lower_x + occupancyMatrixBoxSize/2 , lower_y + occupancyMatrixBoxSize/2 );
+                final Vec2 testpoint2 = new Vec2(lower_x + occupancyMatrixBoxSize/4, lower_y + occupancyMatrixBoxSize/4);
+                final Vec2 testpoint3 = new Vec2(lower_x + 3*occupancyMatrixBoxSize/4, lower_y + 3*occupancyMatrixBoxSize/4);
+                final Vec2 testpoint4 = new Vec2(lower_x + occupancyMatrixBoxSize/4, lower_y + 3*occupancyMatrixBoxSize/4);
+                final Vec2 testpoint5 = new Vec2(lower_x + 3*occupancyMatrixBoxSize/4, lower_y + occupancyMatrixBoxSize/4);
+
+
+                final int icopy = i;
+                final int jcopy = j;
+                QueryCallback callback = new QueryCallback() {
+                    @Override
+                    public boolean reportFixture(Fixture fixture) {
+                        if(fixture.testPoint(testpoint1) || fixture.testPoint(testpoint2) || fixture.testPoint(testpoint3) || fixture.testPoint(testpoint4) || fixture.testPoint(testpoint5)) {
+                            om[icopy][jcopy] = 1;
+                        }
+                        return true;
+                    }
+                };
+                world.queryAABB(callback, currentSquare);
+            }
+        }
+        setOccupancyMatrix(om);
```

Comment:
it looks like this will set om before the queries are all finished. You should do a sort of && check on the finishing of all queries before updating it- or even do something blocking if the method requires that the matrix is filled upon completion. lmk if you need help with that.